testcases/InstallUpstreamKernel.py

```diff
@@ -0,0 +1,112 @@
+        if not (self.hostip and self.user and self.password):
+            self.fail("Provide host ip user details refer, --host-{ip,user,password}")
+
+    def runTest(self):
+        if self.disk:
+            log.debug("Set given disk as default bootable disk")
+            self.cv_SYSTEM.goto_state(OpSystemState.PETITBOOT_SHELL)
+            self.cv_SYSTEM.sys_set_bootdev_no_override()
+            time.sleep(60)
+            cmd = 'blkid %s*' % self.disk
+            output = self.cv_SYSTEM.console.run_command(cmd)
+            uuid = output[0].split(':')[1].split('=')[1].replace("\"", "")
+            cmd = 'nvram --update-config "auto-boot?=true"'
+            output = self.cv_SYSTEM.console.run_command(cmd)
+            cmd = 'nvram --update-config petitboot,bootdevs=uuid:%s' % uuid
+            output = self.cv_SYSTEM.console.run_command(cmd)
+            cmd = 'nvram --print-config'
+            output = self.cv_SYSTEM.console.run_command(cmd)
+        self.cv_SYSTEM.goto_state(OpSystemState.OS)
+
+        try:
+            con = self.cv_SYSTEM.cv_HOST.get_ssh_connection()
+        except Exception:
+            con = self.cv_SYSTEM.console
+
+        try:
+            onlinecpus = int(con.run_command("lscpu --online -e|wc -l")[-1])
+        except Exception:
+            onlinecpus = 20
+        log.debug("Downloading linux kernel")
+        # Compile and install given kernel
+        con.run_command("[ -d %s ] || mkdir -p %s" % (self.home, self.home))
+        con.run_command("if [ -d %s/linux ];then rm -rf %s/linux;fi" % (self.home, self.home))
+        con.run_command("cd %s && git clone --depth 1  %s -b %s linux" % (self.home, self.repo, self.branch), timeout=self.host_cmd_timeout)
+        con.run_command("cd %s/linux" % self.home)
+        if self.patch:
+            log.debug("Applying given patch")
+            # TODO: make copy_test_file_to_host generic to take from any source directory
+            self.cv_HOST.copy_test_file_to_host(self.patch)
+            con.run_command("git am /tmp/%s" % self.patch)
+        log.debug("Downloading linux kernel config")
+        if self.config_path:
+            con.run_command("wget %s -O .config" % self.config_path)
+        con.run_command("make %s" % self.config)
+        log.debug("Compile and install linux kernel")
+        con.run_command("make -j %d -s && make modules && make modules_install && make install" % onlinecpus, timeout=self.host_cmd_timeout)
+        # FIXME: Handle distributions which do not support grub
+        con.run_command("grub2-mkconfig  >/boot/grub2/grub.cfg")
+        con.run_command('grubby --set-default /boot/vmlinuz-*`git describe --always`')
+        log.debug("Rebooting after kernel install...")
+        self.cv_SYSTEM.goto_state(OpSystemState.OFF)
+        self.cv_SYSTEM.goto_state(OpSystemState.OS)
+        try:
+            con = self.cv_SYSTEM.cv_HOST.get_ssh_connection()
+        except Exception:
+            con = self.cv_SYSTEM.console
```

Comment:
Here also SSH connection should suffice i guess? perhaps it needs sometimes here as you are upgrading/installing upstream kernel, ssh connection may fail.